File: lever/src/main/java/com/pin/lever/ContentStreamRequestHandler.java
package com.pin.lever;

import android.content.ContentResolver;
import android.content.Context;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import okio.Okio;
import okio.Source;

import static android.content.ContentResolver.SCHEME_CONTENT;
import static com.pin.lever.Lever.LoadedFrom.DISK;

class ContentStreamRequestHandler extends RequestHandler {
    final Context context;

    ContentStreamRequestHandler(Context context) {
        this.context = context;
    }

    @Override public boolean canHandleRequest(Request data) {
        return SCHEME_CONTENT.equals(data.uri.getScheme());
    }

    @Override public Result load(Request request, int networkPolicy) throws IOException {
        Source source = Okio.source(getInputStream(request));
        return new Result(source, DISK);
    }

    InputStream getInputStream(Request request) throws FileNotFoundException {
        ContentResolver contentResolver = context.getContentResolver();
        return contentResolver.openInputStream(request.uri);
    }
}

File: lever/src/main/java/com/pin/lever/ContentStreamRequestHandler.kt
package com.pin.lever

import android.content.ContentResolver
import android.content.Context
import com.pin.lever.Lever.LoadedFrom
import okio.Okio
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream

internal open class ContentStreamRequestHandler(val context: Context) :
    RequestHandler() {
    override fun canHandleRequest(data: Request?): Boolean {
        return ContentResolver.SCHEME_CONTENT == data!!.uri!!.scheme
    }

    @Throws(IOException::class)
    override fun load(
        request: Request?,
        networkPolicy: Int
    ): Result? {
        val source = Okio.source(getInputStream(request))
        return Result(source, LoadedFrom.DISK)
    }

    @Throws(FileNotFoundException::class)
    fun getInputStream(request: Request?): InputStream {
        val contentResolver = context.contentResolver
        return contentResolver.openInputStream(request!!.uri!!)
    }

}